openslides/assignments/views.py

```diff
@@ -242,6 +243,134 @@ def get_serializer_class(self):
             serializer_class = AssignmentShortSerializer
         return serializer_class
 
+    @detail_route(methods=['post', 'delete'])
+    def candidature_self(self, request, pk=None):
+        """
+        View to nominate self as candidate (POST) or withdraw own candidature
+        (DELETE).
+        """
+        if not request.user.has_perm('assignments.can_nominate_self'):
+            self.permission_denied(request)
+        assignment = self.get_object()
+        if assignment.is_elected(request.user):
+            raise ValidationError({'detail': _('You are already elected.')})
+        if request.method == 'POST':
+            message = self.nominate_self(request, assignment)
+        else:
+            # request.method == 'DELETE'
+            message = self.withdraw_self(request, assignment)
+        return Response({'detail': message})
+
+    def nominate_self(self, request, assignment):
+        if assignment.phase == assignment.PHASE_FINISHED:
+            raise ValidationError({'detail': _('You can not candidate to this election because it is finished.')})
+        if assignment.phase == assignment.PHASE_VOTING and not request.user.has_perm('assignments.can_manage'):
+            # To nominate self during voting you have to be a manager.
+            self.permission_denied(request)
+        if assignment.is_candidate(request.user):
+            raise ValidationError({'detail': _('You are already a candidate of this election.')})
+        assignment.set_candidate(request.user)
+        return _('You were nominated successfully.')
+
+    def withdraw_self(self, request, assignment):
+        # Withdraw candidature and set self blocked.
+        if assignment.phase == assignment.PHASE_FINISHED:
+            raise ValidationError({'detail': _('You can not withdraw your candidature to this election because it is finished.')})
+        if assignment.phase == assignment.PHASE_VOTING and not request.user.has_perm('assignments.can_manage'):
+            # To withdraw self during voting you have to be a manager.
+            self.permission_denied(request)
+        if not assignment.is_candidate(request.user):
+            raise ValidationError({'detail': _('You are not a candidate of this election.')})
+        assignment.set_blocked(request.user)
+        return _(
+            'You have withdrawn your candidature successfully. '
+            'You can not be nominated by other participants anymore.')
+
+    def get_other_user(self, request):
```

Comment:
I don't like this neither. If we have a new case where we need this method in the future, do we have to rename it to "get_user_for_candidature_and_election_and_new_case"? A method should be named what it does and not where it is used.